### DISCUSSION TITLE
How to prevent installing a release in an environment in the wrong kubecontext

### DISCUSSION THREAD

**Discussion by cjroebuck** [helmfile/helmfile#1709] [Questions]
Let's say I have 

```
environments:
  prod:
    values:
      domain: prod.example.com
  
  staging:
    values:
      domain: staging.example.com

```

and some release, called `api-release` which looks up `.Values.domain`

I also have kubecontext's for a `prod-cluster` and `staging-cluster`. 

Under no circumstances should `helmfile -l name=api-release -e staging apply` ever be able to run when my local kubecontext is set to `prod-cluster`, but this can easily happen if we're not careful about which kubecontext is currently set locally.

I had thought that setting `kubeContext` alongside the environments would work, but it doesn't, so are there any other ways, for e.g. using a global hook perhaps? 

I should say I'm also using sub-helmfiles and passing down the env values using: `- {{ toYaml .Values | nindent 6 }}` as explained in https://helmfile.readthedocs.io/en/latest/writing-helmfile/#re-using-environment-state-in-sub-helmfiles

**Comment by yxxhero** [comment]
@cjroebuck pls post an example.

**Comment by GideonStowell** [comment]
Not to hijack but I have the same question and this is what I am working with:

helmfile.yaml
```
environments:
  development:
    kubeContext: eks-dev-uswest2

repositories:
- name: airbyte
  url: https://airbytehq.github.io/helm-charts
---
releases:
- name: dev-airbyte
  namespace: airbyte
  chart: airbyte/airbyte
  version: 0.493.1
  values:
  - ./airbyte/development/values.yml
  kubeContext: "{{.Environment.KubeContext}}"
```

```
$ helmfile apply -f helmfile.development.yml --environment development
Adding repo airbyte https://airbytehq.github.io/helm-charts
"airbyte" has been added to your repositories

Comparing release=dev-airbyte, chart=airbyte/airbyte, namespace=airbyte
in ./helmfile.development.yml: command "/home/linuxbrew/.linuxbrew/bin/helm" exited with non-zero status:

PATH:
  /home/linuxbrew/.linuxbrew/bin/helm

ARGS:
  0: helm (4 bytes)
  1: diff (4 bytes)
  2: upgrade (7 bytes)
  3: --allow-unreleased (18 bytes)
  4: dev-airbyte (11 bytes)
  5: airbyte/airbyte (15 bytes)
  6: --version (9 bytes)
  7: 0.493.1 (7 bytes)
  8: --kube-context (14 bytes)
  9: {{.Environment.KubeContext}} (28 bytes)
  10: --namespace (11 bytes)
  11: airbyte (7 bytes)
  12: --values (8 bytes)
  13: /tmp/helmfile2364251420/airbyte-dev-airbyte-values-c49fcd5d7 (60 bytes)
  14: --detailed-exitcode (19 bytes)
  15: --color (7 bytes)
  16: --reset-values (14 bytes)

ERROR:
  exit status 1

EXIT STATUS
  1

STDERR:
  Error: Failed to get release dev-airbyte in namespace airbyte: exit status 1: Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist
  Error: plugin "diff" exited with error

COMBINED OUTPUT:
  Error: Failed to get release dev-airbyte in namespace airbyte: exit status 1: Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist
  Error: plugin "diff" exited with error
```

The `eks-dev-uswest2` context does exist:
```
$ kubectl config get-contexts
CURRENT   NAME                CLUSTER                                                        AUTHINFO                                                       NAMESPACE
*         eks-dev-uswest2   arn:aws:eks:us-west-2:XXXXXX:cluster/eks-dev-uswest2   arn:aws:eks:us-west-2:XXXXXX:cluster/eks-dev-uswest2
*
```

To me it seems like the `.Environment.KubeContext` is not being rendered for some reason. I get similar output when I use other helmfile commands beside `apply`, btw.

@yxxhero Any ideas on what I am doing wrong?

**Reply by yxxhero** [reply]
Please set your helmfile endwith tmpl.

**Reply by GideonStowell** [reply]
@yxxhero you mean just change the file ending to be tmpl instead of yml/yaml? 

I tried this and it did not change the behavior. Can you perhaps show an example of it would be done if I am missing something?

**Reply by yxxhero** [reply]
@GideonStowell please show the log.

**Reply by GideonStowell** [reply]
@yxxhero 
When I said it didn't change the behavior I mean it had the exact same behavior has before I changed the file ending:

```
$ helmfile status -f helmfile.development.tmpl --environment development
Getting status dev-airbyte
in ./helmfile.development.tmpl: release "dev-airbyte" failed: command "/home/linuxbrew/.linuxbrew/bin/helm" exited with non-zero status:

PATH:
  /home/linuxbrew/.linuxbrew/bin/helm

ARGS:
  0: helm (4 bytes)
  1: status (6 bytes)
  2: dev-airbyte (11 bytes)
  3: --namespace (11 bytes)
  4: airbyte (7 bytes)
  5: --kube-context (14 bytes)
  6: {{.Environment.KubeContext}} (28 bytes)

ERROR:
  exit status 1

EXIT STATUS
  1

STDERR:
  Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist

COMBINED OUTPUT:
  Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist
```

The same error occurs: `Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist`. It still indicates to me that the template isn't getting rendered.

**Comment by cjroebuck** [comment]
The way I solved this was using a separate file i.e. `kubecontexts.yaml` with all of my environments listed, along with the kubeContext next to them. 

Then in my sub-helmfiles, I simply include this file as a 'base':

```
bases: 
  - ../kubecontexts.yaml
```

Now the kubecontext does get picked up whenever I switch environments.

**Reply by GideonStowell** [reply]
Can you share your helmfile.yml and bases.yml files?

Mine look like this but I still get errors when running any helmfile commands:

```
# bases.yml
environments:
  development:
    kubeContext: dev-uswest2
```

```
# helmfile.yml
bases:
  - bases.yml

repositories:
- name: airbyte
  url: https://airbytehq.github.io/helm-charts

releases:
- name: dev-airbyte
  namespace: airbyte
  chart: airbyte/airbyte
  version: 0.493.1
  values:
  - ./airbyte/development/values.yml
  kubeContext: "{{.Environment.KubeContext}}"
```
And this is the error that I am getting, as you should be able to see, it is not rendering the `.Environment.KubeContext` variable:

```
$ helmfile status -f helmfile.development.yml --environment development
Getting status dev-airbyte
in ./helmfile.development.yml: release "dev-airbyte" failed: command "/home/linuxbrew/.linuxbrew/bin/helm" exited with non-zero status:

PATH:
  /home/linuxbrew/.linuxbrew/bin/helm

ARGS:
  0: helm (4 bytes)
  1: status (6 bytes)
  2: dev-airbyte (11 bytes)
  3: --namespace (11 bytes)
  4: airbyte (7 bytes)
  5: --kube-context (14 bytes)
  6: {{.Environment.KubeContext}} (28 bytes)

ERROR:
  exit status 1

EXIT STATUS
  1

STDERR:
  Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist

COMBINED OUTPUT:
  Error: Kubernetes cluster unreachable: context "{{.Environment.KubeContext}}" does not exist
```

**Reply by cjroebuck** [reply]
I may be misunderstanding but I don’t think you need to re-set the kubecontext from the environment on your release. 

What happens if you just remove that line from your release? it should just work, as you’ve already set the kubecontext in the environments file and you’re specifying the environment in your command.

**Reply by GideonStowell** [reply]
You are right. I don't need to set `kubeContect` on the individual release. 
I saw that somewhere in the docs as a valid argument but I guess when using the environment `kubeContext` it is not required.
Thanks for the help!
